internal/vpn.go

```diff
@@ -115,8 +115,18 @@ func updateConfig(app *App, client *http.Client, pubkey string) error {
 func initVpn(app *App, client *http.Client, crypto CryptoHandler) error {
 	sotaConfig := filepath.Dir(app.EncryptedConfig)
 	wgPriv := filepath.Join(sotaConfig, "wg-priv")
+	register := false
 	if _, err := os.Stat(wgPriv); os.IsNotExist(err) {
+		register = true
 		log.Println("Wireguard private key does not exist, generating.")
+	} else {
+		secretFile := filepath.Join(app.SecretsDir, "wireguard-client")
+		if _, err := os.Stat(secretFile); os.IsNotExist(err) {
+			register = true
+			log.Println("Wireguard key not registered on server, will re-register")
+		}
```

Comment:
This seems to be dangerous to me and may lead to the following race condition:
1. A fioconfig checks-in and notices there is no wg-priv, so it registers to upload its key (by creating the client-config);
2. The upload is successful (which means the config was created), but for some reason fioconfig fails to download it (e.g. due to the timing issues between device-gateway, CouchBase, and fioconfig itself, or network error).
3. On the next check-in fioconfig will notice there is no client-config and try to register again, although it was already registered
4. In this case a real situation might be that the device receives the client-config from step 2 while it generated a new key at step 3
   Obviously, this condition is temporary and fioconfig will manage to download the new client-config after some time; but that "some time" might mean a quite long period in time.

A morale: while trying to fix a "theoretical" problem (upload of the `client-config` succeeded but something wen wrong on the server-side) we end up creating another "theoretical" problem (a race condition between the server and the client when uploading/downloading the `client-config`).

Reply:
So IMHO, the first commit is enough to fix a real problem.
The second one tries to fix the synthetic problem, but I don't see how it could happen in reality.

The only real use case I can imagine is that the device uploaded `wireguard-client`, but the user decided to delete that config file (e.g. using `fioconfig device config delete <dev> wireguard-client`).
Although that might be a user mistake, it could also be a legitimate use case, so I'd hesitate to re-create this file (effectively overriding an intentional user action).

Reply:
it's not theoretical. a customer's device is stuck right now in this state (priv key exists, but never got uploaded). it would be really nice to determine a way that an OTA would resolve that.

Reply:
I suspected the answer could be like this.
And I doubt I would find a better solution for this given all the constraints.

Reply:
i just force-pushed https://github.com/foundriesio/fioconfig/pull/21/commits/e6634280521f948cc962b6ed0082d77b64e16e5d i think its a little better - and also will be easier to cleanly remove from code some day

Reply:
wow, you've did it.
I like this new way so much 🥇